### PR TITLE
runtime: add DAN support for VFIO network device in Go kata-runtime

### PR DESCRIPTION
**Pull request by l8huang**
When using network adapters that support SR-IOV, a VFIO device can be
plugged into a guest VM and claimed as a network interface. This can
significantly enhance network performance.

Fixes: #9758

### CODE REVIEW

**Comment by l8huang** [comment]
I'm uncertain when will https://github.com/kata-containers/kata-containers/pull/9868 be merged, so I've created this PR to initiate the review process earlier.

@justxuewei could you please take a look at the 2nd commit in this PR? Thanks

**Comment by justxuewei** [comment]
> > I'm uncertain when will #9868 be merged, so I've created this PR to initiate the review process earlier.
> > @justxuewei could you please take a look at the 2nd commit in this PR? Thanks
> 
> I'd like confirm that, the first commit in this PR equals to the [PR#9868](https://github.com/kata-containers/kata-containers/pull/9868) , right ?

Yeah. Once the first PR get merged, this PR only contains one commit.

**Comment by Apokleos** [comment]
> > > I'm uncertain when will #9868 be merged, so I've created this PR to initiate the review process earlier.
> > > @justxuewei could you please take a look at the 2nd commit in this PR? Thanks
> > 
> > 
> > I'd like confirm that, the first commit in this PR equals to the [PR#9868](https://github.com/kata-containers/kata-containers/pull/9868) , right ?
> 
> Yeah. Once the first PR gets merged, this PR only contains one commit.
>
Ok, Let's push the first one first

**Comment by l8huang** [comment]
@justxuewei did you get a chance to review this PR?

**Comment by justxuewei** [comment]
> @justxuewei @Apokleos   did you get a chance to review this PR?

I'm on vacation, so I'll review this PR next week. Is it ok to you?

**Comment by l8huang** [comment]
> I'm on vacation, so I'll review this PR next week. Is it ok to you?

sure, have a good vacation, thanks.

File: src/runtime/virtcontainers/persist/api/network.go
Comment (amshinde) [inline src/runtime/virtcontainers/persist/api/network.go:97]: @l8huang We currently have `PhysicalEndpoint` to support and detect SRIOV network interface or an entire physical network interface. 
Is that different from your use-case?

Comment (l8huang) [reply src/runtime/virtcontainers/persist/api/network.go:97]: PhysicalEndpoint is for VF device, which presents a network interface on host; this PR is for VFIO passthrough device, on host it doesn't has a network interface, it supposes to be plugged into guest VM and recognized by NIC driver in guest kernel to create its network interface.

**Comment by l8huang** [comment]
@justxuewei did you get a chance to review this PR? hope you had a great vacation.

**Comment by justxuewei** [comment]
> @justxuewei did you get a chance to review this PR? hope you had a great vacation.

Thanks. I'll take a look this week ;)

**Comment by justxuewei** [comment]
Could you take a loot at this PR? @Apokleos

File: src/runtime/virtcontainers/network.go
Comment (justxuewei) [inline src/runtime/virtcontainers/network.go:239]: I think you cannot just simply remove it, because this changes the original logic. I’d like to suggest adding a new argument, `enable_netns bool`. If VFIO is enabled, or in some scenarios where you don't want to use netns, set the value to `false`.

```
if enable_netns && network.NetworkID() == "" {
	return nil, nil, nil, nil
}
```

**Review by justxuewei — COMMENTED** [review]
Lgtm, but two nits.

File: src/runtime/virtcontainers/kata_agent.go
Comment (l8huang) [reply src/runtime/virtcontainers/kata_agent.go]: updated

File: src/runtime/virtcontainers/network.go
Comment (l8huang) [reply src/runtime/virtcontainers/network.go:239]: good point. Because `generateVCNetworkStructures()` is wrapped by `setupNetworks()` function now, I added the old check back to `setupNetworks()`.

**Review by l8huang — COMMENTED** [review]
@justxuewei  thanks for reviewing this PR, I updated it according to your comments, PTAL.

File: src/runtime/virtcontainers/kata_agent.go
Comment (l8huang) [inline src/runtime/virtcontainers/kata_agent.go:1243]: @justxuewei the check you mentioned is added at here.

File: src/runtime/virtcontainers/kata_agent.go
Comment (justxuewei) [inline src/runtime/virtcontainers/kata_agent.go:1267]: Are any rollback measures needed here? Assuming that there are 4 VFIO endpoints.

1. Added 2 VFIO eps.
2. Some errors happen at 3 ep.
3. Rollback the above 2 eps, otherwise they might be added twice in the next call to `setupNetworks`.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks!

**Comment by l8huang** [comment]
> Lgtm, thanks!

Cool. @Apokleos @lifupan could you please take 2nd look? thanks.

File: src/runtime/virtcontainers/endpoint.go
Comment (amshinde) [inline src/runtime/virtcontainers/endpoint.go:70]: The name and description here is quite confusing with `PhysicalEndpointType`.  `PhysicalEndpointType` was meant to represent an entire physical device or SRIOV interface that could be passed through VFIO. 
To clearly distinguish between the two, I suggest we should consider changing the naming here, maybe rename this to `DanEndPointType` or `DanVfioEndpointType`.
I also suggest to change the documentation comments to explicitly mention Direct attached networks. Without this, it will be a tall task for future developers to distinguish between the two endpoints.

Comment (amshinde) [reply src/runtime/virtcontainers/endpoint.go:70]: I will take a detailed look at this PR by tomorrow to provide any feedback.

Comment (l8huang) [reply src/runtime/virtcontainers/endpoint.go:70]: Thanks for looking at this PR.

If the name `VfioEndpointType` causing confusion, **I'm okay with `DanVfioEndpointType`** -- `DanEndPointType` looks like too generic. @justxuewei are you okay with `DanVfioEndpointType`? 

BTW: I saw `PhysicalEndpointType` unbinds VF device from its driver then rebind it to `vfio-pci` and cold-plug it to VM. When using ovn-kubernetes, the VF and VFIO devices are pre-allocated on the host, based on the network config on a workload(VM/Pod), a VF or VFIO device can be assigned to it accordingly -- a VFIO  device can be directly assigned to a VM workload, so no need to allocate and config a VF device then use PhysicalEndpoint to unbind/rebind it as VFIO, see https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/cni/helper_linux.go#L293-L301. 

One question out of curiosity: In the use case of PhysicalEndpoint, which CNI is typically utilized?

Comment (l8huang) [reply src/runtime/virtcontainers/endpoint.go:70]: @amshinde do you have further comments for this PR?

Comment (l8huang) [reply src/runtime/virtcontainers/endpoint.go:74]: @amshinde I added a doc string to clarify the difference from PhysicalEndpointType. I also renaming,  but since the endpoint type is inherently VFIO and DAN is merely the configuration method, renaming it to DanVfioEndpointType might introduce unnecessary confusion. Even if we renamed it, the value would still be 'vfio'. Therefore, I prefer to keep the current name. Does that work for you?"

Comment (justxuewei) [reply src/runtime/virtcontainers/endpoint.go:74]: I don't think adding DAN is a good idea, since DAN stands for Directy Attachable Network. It just a new way of reading network config, and is not relevant for endpoints.

Comment (l8huang) [reply src/runtime/virtcontainers/endpoint.go:74]: @amshinde do you have any further feedback on this PR?

Comment (amshinde) [reply src/runtime/virtcontainers/endpoint.go:74]: > One question out of curiosity: In the use case of PhysicalEndpoint, which CNI is typically utilized?

We used this with sriov-device-network-plugin : https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin
The plugin creates a VF and places it in the network namespace of the sandbox.

Comment (amshinde) [reply src/runtime/virtcontainers/endpoint.go:74]: got delayed a bit, I'll take another look and provide my review by today.

**Comment by amshinde** [comment]
@l8huang I went over this PR, and I am not really clear about the use case here.
Can you provide an explanation of the end to end flow, what CNI is being used.

We currently have `PhysicalEndpoint` which is used in the use-case for `sriov-device-network-plugin` that places the VF in the network namespace of the container. The kata-runtime then scans the network namespace to check if is this is physical device, binds it to vfio instead and does a device passthrough.

Looks like in your case the binding to VFIO is done by CNI plugin itself? (Curious why the above flow would not work for you). My question to you then is how is the device passed to the runtime? You mentioned a KEP in the PR, but as far as I understand, the KEP is meant to pass the information of all devices beforehand to account for things like pcie address space required. 

Can you please clarify the above and include the usage in the PR/commit as well. Eventually it will be useful to have a doc including  a detailed explanation here: https://github.com/kata-containers/kata-containers/tree/main/docs/use-cases

File: src/runtime/virtcontainers/vfio_endpoint.go
Comment (amshinde) [inline src/runtime/virtcontainers/vfio_endpoint.go:69]: Why is Attach/Detach not implemented at all?

File: src/runtime/virtcontainers/sandbox.go
Comment (amshinde) [inline src/runtime/virtcontainers/sandbox.go:2241]: Seems like an odd place to put this function here. Consider adding this to https://github.com/kata-containers/kata-containers/blob/main/src/runtime/pkg/device/drivers/vfio.go

File: src/runtime/virtcontainers/kata_agent.go
Comment (amshinde) [inline src/runtime/virtcontainers/kata_agent.go:1264]: /BFD/BDF/

File: src/runtime/virtcontainers/vfio_endpoint.go
Comment (l8huang) [reply src/runtime/virtcontainers/vfio_endpoint.go:69]: VFIO device is attached/detached by VFIODevice's Attach()/Detach() function.

File: src/runtime/virtcontainers/sandbox.go
Comment (l8huang) [reply src/runtime/virtcontainers/sandbox.go:2241]: [src/runtime/pkg/device/drivers/vfio.go](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/pkg/device/drivers/vfio.go) is for a single VFIODevice. This function needs to iterate all `VFIODev`s **via `Sandbox.devManager`** to find the `VFIODev` which matches `hostBDF`, IMO it's not reasonable to move this function to vfio.go.

File: src/runtime/virtcontainers/kata_agent.go
Comment (l8huang) [reply src/runtime/virtcontainers/kata_agent.go:1264]: will do.

**Comment by l8huang** [comment]
> We currently have `PhysicalEndpoint` which is used in the use-case for `sriov-device-network-plugin` that places the VF in the network namespace of the container. The kata-runtime then scans the network namespace to check if is this is physical device, binds it to vfio instead and does a device passthrough.
>
> Looks like in your case the binding to VFIO is done by CNI plugin itself? (Curious why the above flow would not work for you). My question to you then is how is the device passed to the runtime? You mentioned a KEP in the PR, but as far as I understand, the KEP is meant to pass the information of all devices beforehand to account for things like pcie address space required.

Looks like you missed the related discussions. Long story short, In our use case, we uses `sriov-device-network-plugin` to allocate VFs(for regular k8s Pod) and VFIOs(for KubeVirt VM), **the number of them are fixed based on the role of k8s node, once they are allocated we don't unbind/bind them in runtime any more**. Our CNI is ovn-k8s; for regular k8s Pod one or more VFs will be allocated and ovn-k8s configures the networks on them; for KubeVirt VM, VFIOs will be allocated, ovn-k8s only adds their representers to OVS bridge and provides network configuration to VM by annotation `k8s.v1.cni.cncf.io/network-status`(ref: [Kubernetes Network Custom Resource Definition De-facto Standard](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit)) -- **it's not necessary to configure a VF interface with network/route/etc. and rely on VM runtime to scan the config, unbind/rebind the device**.

With Kata, DAN is a neat solution for VFIO network interface, ovn-k8s configures VFIO representers as it currently does, then generates a DAN configuration to apply network setting on the network interface claimed by guest kernel. On the host node, no bind/unbind operations are required, making it a more straightforward approach to use VFIO network device.

The KEP is for moving the `resources` from container-level to Pod lever, so the VFIO device allocation can happens before creating guest VM, so they can be cold-plugged. The VFIO network device's PCI topology is critical for GPUDirect RDMA, per my understanding, GPU/DPU will use CDI to achieve the desired topology. CC @zvonkok

**Comment by l8huang** [comment]
@amshinde thanks for your comments, PR updated, PTAL

I will add a doc to `docs/use-cases` later.

**Comment by l8huang** [comment]
P.S. our sriov-network-device plugin uses config like below to allocate VF and VFIO devices on network adapter:

```
 {
    "skipTopology": true,
    "resourceList": [
        {
            "resourceName": "vf",
            "resourcePrefix": "nvidia.com",
            "selectors": {
                "vendors": [ "15b3" ],
                "devices": [ "101e" ],
                "pfNames": [ "enp1s0f0#1-8" ] # 8 VFs
            }
        },
        {
            "resourceName": "vfio",
            "resourcePrefix": "nvidia.com",
            "selectors": {
                "vendors": [ "15b3" ],
                "devices": [ "101e" ],
                "drivers": [ "vfio-pci" ],
                "pfNames": [ "enp1s0f0#9-16" ] # 8 VFIOS
            }
        }
    ]
}
```
And ovn-k8s uses annotation `k8s.v1.cni.cncf.io/resourceName` to specify should VF or VFIO to be allocated for a network: 
```
- apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      k8s.v1.cni.cncf.io/resourceName: nvidia.com/vf # allocate a VF device, ovn-k8s config its network interface
    labels:
    name: ovn-primary
...
- apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      k8s.v1.cni.cncf.io/resourceName: nvidia.com/vfio # allocate a VFIo device, ovn-k8s uses DAN to config network interface inside guest VM
    labels:
    name: ovn-secondary-vfio
```

@amshinde please let me know if you have further questions.

**Comment by l8huang** [comment]
@Apokleos @lifupan could you please help review this PR?

**Comment by l8huang** [comment]
This PR has been awaiting a second approval for a month. Please kindly let me know how to move it forward?

**Review by Apokleos — COMMENTED** [review]
Thx @l8huang for your work on DAN for runtime-go. I've gone over the implementation and left two comments on the parts that confused me. For the most part, looks good to me.

File: src/runtime/virtcontainers/network_linux.go
Comment (l8huang) [reply src/runtime/virtcontainers/network_linux.go:465]: `DanDevice` defines the generic configuration of a network interface(ip/route/neigh), while `VfioDanDevice` is a specific type of device that is configured using DAN.

runtime-rs supports other types of device:

```
pub(crate) enum Device {
    #[serde(rename = "vhost-user")]
    VhostUser {
        // Vhost-user socket path
        path: String,
        #[serde(default)]
        queue_num: usize,
        #[serde(default)]
        queue_size: usize,
    },
    #[serde(rename = "host-tap")]
    HostTap {
        tap_name: String,
        #[serde(default)]
        queue_num: usize,
        #[serde(default)]
        queue_size: usize,
    },
}
```

see  https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/resource/src/network/dan.rs#L237-L255

File: src/runtime/virtcontainers/sandbox.go
Comment (l8huang) [reply src/runtime/virtcontainers/sandbox.go:252]: it's a flag default to false, once hot-plugged network interface(VFIO in this case)'s network config applied successfully, it will be set to true. It's for preventing network config API being called multiple times when Pod has multiple containers.

PTAL comment at https://github.com/kata-containers/kata-containers/pull/9977/files#diff-6fb7f551035031ed29ad100c61f1a7e92b0fa15d2f67e4c7ea7ebf3b715583f4R1249-R1261

Comment (l8huang) [reply src/runtime/virtcontainers/sandbox.go:252]: @Apokleos thanks for reviewing this PR

File: src/runtime/virtcontainers/network_linux.go
Comment (Apokleos) [reply src/runtime/virtcontainers/network_linux.go:465]: Thx @l8huang do you have plan to support other types of devices listed above

File: src/runtime/virtcontainers/sandbox.go
Comment (Apokleos) [reply src/runtime/virtcontainers/sandbox.go:252]: Could you please add comment to explain this introduced flag ?

Comment (Apokleos) [reply src/runtime/virtcontainers/sandbox.go:252]: This PR almost looks good to me but the two comments !

File: src/runtime/virtcontainers/network_linux.go
Comment (l8huang) [reply src/runtime/virtcontainers/network_linux.go:465]: nope, we don't have use cases for `vhost-user` and `host-tap`.

File: src/runtime/virtcontainers/sandbox.go
Comment (l8huang) [reply src/runtime/virtcontainers/sandbox.go:255]: added docstring for the flag, PTAL.

**Review by Apokleos — APPROVED** [review]
thx @l8huang LGTM ！

**Pull request merged**
